src/impression.js

```diff
@@ -53,74 +53,109 @@ export function resetTrackImpressionPromiseForTesting() {
+      resolveImpression();
+    }, () => {});
   });
 }
 
+/**
+ * Signal that impression tracking is not relevant in this environment.
+ */
+export function doNotTrackImpression() {
+  trackImpressionPromise = Promise.resolve();
+}
 
 /**
- * Actually perform the impression request if it has been provided via
- * the click param in the viewer arguments.
+ * Handle the getReplaceUrl and return a promise when url is replaced
+ * Only handles replaceUrl when viewer indicates AMP to do so. Viewer should indicate
+ * by setting the legacy replaceUrl init param and add `replaceUrl` to its capability param.
+ * Future plan is to change the type of legacy init replaceUrl param from url string
+ * to boolean value.
+ * Please NOTE replaceUrl and adLocation will never arrive at same time,
+ * so there is no race condition on the order of handling url replacement.
  * @param {!Window} win
- * @param {!./service/viewer-impl.Viewer} viewer
- * @param {!Function} resolveImpression Call when the impression has been
- *     tracked.
+ * @return {!Promise}
  */
-function trackImpressionFromClickParam(win, viewer, resolveImpression) {
+function handleReplaceUrl(win) {
+  const viewer = Services.viewerForDoc(win.document);
+  if (!viewer.getParam('replaceUrl') || !viewer.hasCapability('replaceUrl')) {
+    // Right now we use the legacy replaceUrl init param as a signal to request
+    // replaceUrl from viewer.
+    // Viewer's capability to support getting replaceUrl is also required.
+    return Promise.resolve();
+  }
+  return viewer.whenFirstVisible().then(() => {
+    viewer.sendMessageAwaitResponse('getReplaceUrl', undefined).then(
+        response => {
+          if (!response || typeof response != 'object') {
+            dev().warn('IMPRESSION', 'get invalid replaceUrl response');
+            return;
+          }
+          if (response['replaceUrl']) {
+            viewer.replaceUrl(response['replaceUrl']);
+          }
+        }, () => {});
+  });
+}
+
+
+/**
+ * Perform the impression request if it has been provided via
+ * the click param in the viewer arguments. Returns a promise.
+ * @param {!Window} win
+ * @return {!Promise}
+ */
+function handleClickUrl(win) {
+  const viewer = Services.viewerForDoc(win.document);
   /** @const {string|undefined} */
   const clickUrl = viewer.getParam('click');
 
   if (!clickUrl) {
-    resolveImpression();
-    return;
+    return Promise.resolve();
   }
+
   if (clickUrl.indexOf('https://') != 0) {
     user().warn('IMPRESSION',
         'click fragment param should start with https://. Found ',
         clickUrl);
-    resolveImpression();
-    return;
+    return Promise.resolve();
   }
+
   if (win.location.hash) {
     // This is typically done using replaceState inside the viewer.
     // If for some reason it failed, get rid of the fragment here to
     // avoid duplicate tracking.
     win.location.hash = '';
   }
 
-  viewer.whenFirstVisible().then(() => {
```

Comment:
clickUrl should only be invoked when the document becomes visible (it ends up logging the click, which we shouldn't do unless the document is viewed)

Reply:
oops, good catch! fixed.